Source/WebCore/Modules/webaudio/AsyncAudioDecoder.cpp

```diff
@@ -29,86 +29,27 @@
 #include "AsyncAudioDecoder.h"
 
 #include "AudioBuffer.h"
-#include "AudioBufferCallback.h"
-#include "ExceptionOr.h"
 #include <JavaScriptCore/GenericTypedArrayViewInlines.h>
 #include <JavaScriptCore/TypedArrayAdaptors.h>
-#include <wtf/MainThread.h>
+#include <wtf/NativePromise.h>
 
 namespace WebCore {
 
 AsyncAudioDecoder::AsyncAudioDecoder()
+    : m_runLoop(RunLoop::create("Audio Decoder", ThreadType::Audio))
 {
-    // Start worker thread.
-    Locker locker { m_threadCreationMutex };
-    m_thread = Thread::create("Audio Decoder", [this] {
-        runLoop();
-    }, ThreadType::Audio);
 }
 
-AsyncAudioDecoder::~AsyncAudioDecoder()
+Ref<DecodingTaskPromise> AsyncAudioDecoder::decodeAsync(Ref<ArrayBuffer>&& audioData, float sampleRate)
 {
-    m_queue.kill();
-    
-    // Stop thread.
-    m_thread->waitForCompletion();
-}
-
-void AsyncAudioDecoder::decodeAsync(Ref<ArrayBuffer>&& audioData, float sampleRate, Function<void(ExceptionOr<Ref<AudioBuffer>>&&)>&& callback)
-{
-    ASSERT(isMainThread());
-
-    auto decodingTask = makeUnique<DecodingTask>(WTFMove(audioData), sampleRate, WTFMove(callback));
-    m_queue.append(WTFMove(decodingTask)); // note that ownership of the task is effectively taken by the queue.
-}
-
-void AsyncAudioDecoder::runLoop()
-{
-    ASSERT(!isMainThread());
-
-    {
-        // Wait for until we have m_thread established before starting the run loop.
-        Locker locker { m_threadCreationMutex };
-    }
-
-    // Keep running decoding tasks until we're killed.
-    while (auto decodingTask = m_queue.waitForMessage()) {
-        // Let the task take care of its own ownership.
-        // See DecodingTask::notifyComplete() for cleanup.
-        decodingTask.release()->decode();
-    }
-}
-
-AsyncAudioDecoder::DecodingTask::DecodingTask(Ref<ArrayBuffer>&& audioData, float sampleRate, Function<void(ExceptionOr<Ref<AudioBuffer>>&&)>&& callback)
-    : m_audioData(WTFMove(audioData))
-    , m_sampleRate(sampleRate)
-    , m_callback(WTFMove(callback))
-{
-}
-
-void AsyncAudioDecoder::DecodingTask::decode()
-{
-    // Do the actual decoding and invoke the callback.
-    m_audioBuffer = AudioBuffer::createFromAudioFileData(m_audioData->data(), m_audioData->byteLength(), false, sampleRate());
-    
-    // Decoding is finished, but we need to do the callbacks on the main thread.
-    callOnMainThread([this] {
-        notifyComplete();
+    return WTF::invokeAsync(m_runLoop, __func__, [audioData = WTFMove(audioData), sampleRate] {
+        auto audioBuffer = AudioBuffer::createFromAudioFileData(audioData->data(), audioData->byteLength(), false, sampleRate);
+        if (!audioBuffer)
+            return DecodingTaskPromise::createAndReject(EncodingError, __func__);
+        return DecodingTaskPromise::createAndResolve(audioBuffer.releaseNonNull(), __func__);
```

Comment:
I updated the documentation in the NativePromise PR. Move will always be used. If the object isn't movable, then it won't compile. Simple as that.
From the documentation.
```
 * When IsExclusive is true:
 *  - The NativePromise performs run-time assertions that there is at most one call to either then(...) or chainTo(...).
 *  - Move semantics are used when passing arguments
 *  - The resolved or rejected object will be deleted on the target thread.
 *  - The ResolveValueType and RejectValueType must be moveable (e.g have a move constructor). Compilation will fail otherwise.
 * Otherwise:
 * - values are passed to the resolve/reject callbacks through either const references or pointers.
 * - the resolve or reject object will be deleted on the last SerialFunctionDispatcher that got used.
```
The entire chain of function parameters when IsExclusive is true is ResolveValueType&&

There's of course the issue if the promise isn't then()ed or whenSettled() then the scope of the value is the same as the promise. But that isn't the case here.

In this particular case, it is obvious what is going on, for the simple case that EWS shows no crash.
AudioBuffer isn't a DOM object, it has a non-thread-safe refcount and it is moved. If there was, RefCounted<AudioBuffer> would have asserted when it goes out of scope on the main thread.

Also, the flow of this AudioBuffer from decoder runLoop back to the main thread is identical to before. Why would 
consider it a problem all of a sudden?
Code before was doing exactly that: allocate on the decoder's runLoop, move it to the main thread and return it to the caller.

> For instance, resolving a promise with a RefCounted object that has a ref count of more than 1 is potentially problematic given caller and callee might not know whether they work on the same thread.

While NativePromise greatly simplify the use and add lots of safety layers, you would expect a WebKit programmer to have some clue on what they are doing when dealing with multi-threading and they decide to share a RefPtr<T> across multiple-threads and use it at the same time !

This change is all about using NativePromise, not arguing again on how NativePromise is written. I had hoped that had been settled by now :(